lib/util/lifted/influx/httpd/handler_metrics.go

```diff
@@ -0,0 +1,259 @@
+	openGeminiCollector *OpenGeminiCollector
+)
+
+func init() {
+	// Create a custom registry
+	registry := prometheus.NewRegistry()
+	// Instantiate a custom collector
+	openGeminiCollector = NewOpenGeminiCollector()
+	// Register the collector to the registry
+	err := registry.Register(openGeminiCollector)
+	if err != nil {
+		log.Fatal("Failed to register collector:", err)
+	}
+
+	metricsHandler = promhttp.HandlerFor(registry, promhttp.HandlerOpts{Registry: registry})
+}
+
+type OpenGeminiCollector struct {
+	*metrics.BaseCollector
+	indexMap map[string]*metrics.ModuleIndex
+}
+
+func NewOpenGeminiCollector() *OpenGeminiCollector {
+	c := &OpenGeminiCollector{
+		BaseCollector: metrics.NewBaseCollector(),
+		indexMap:      make(map[string]*metrics.ModuleIndex),
+	}
+	// Initialize metrics
+	for moduleName, index := range c.IndexRegistry {
+		desc := make(map[string]*prometheus.Desc)
+		for indexName, help := range index.HelpMap {
+			desc[indexName] = metrics.NewDesc("", indexName, help, index.Labels)
+		}
+		c.AllModulesDesc[moduleName] = desc
+	}
+
+	return c
+}
+
+func (c *OpenGeminiCollector) Collect(ch chan<- prometheus.Metric) {
+	var indexValue interface{}
+	indexMap := c.indexMap
+	for moduleName, index := range c.IndexRegistry {
+		value, ok := indexMap[moduleName]
+		labelErr := false
+		if ok {
+			// Get the label of the module
+			labelValues := make([]string, 0)
+			for _, labelName := range index.Labels {
+				labelValue, ok := value.LabelValues[labelName]
+				if !ok {
+					labelErr = true
+					break
+				}
+				labelValues = append(labelValues, labelValue)
+			}
+			if labelErr {
+				continue
+			}
+
+			for indexName := range index.HelpMap {
+				indexValue, ok = value.MetricsMap[indexName]
+				// If the indicator exists, write the indicator
+				if ok {
+					m := prometheus.MustNewConstMetric(c.AllModulesDesc[moduleName][indexName], prometheus.GaugeValue,
+						indexValue.(float64), labelValues...)
+
+					ch <- prometheus.NewMetricWithTimestamp(value.Timestamp, m)
+				}
+			}
+		}
+	}
+
+}
+
+func (h *Handler) serveMetrics(w http.ResponseWriter, r *http.Request, user meta.User) {
+	for moduleName, index := range openGeminiCollector.IndexRegistry {
+		moduleIndex, err := getMetrics(h, r, user, moduleName, index.Labels)
+		if err != nil {
+			continue
+		}
+		openGeminiCollector.indexMap[moduleName] = moduleIndex
+	}
+
+	metricsHandler.ServeHTTP(w, r)
+}
+
+// // getMetrics serverProm
+func getMetrics(h *Handler, r *http.Request, user meta.User, tableName string, labelNames []string) (*metrics.ModuleIndex, error) {
+	nodeID, _ := strconv.ParseUint(r.FormValue("node_id"), 10, 64)
+	var q *influxql.Query
+	var err error
+	sql := fmt.Sprintf("select * from %s", tableName)
+	for i, labelName := range labelNames {
+		if i == 0 {
+			sql = fmt.Sprintf("%s group by %s", sql, labelName)
+		} else {
+			sql = fmt.Sprintf("%s,%s", sql, labelName)
+		}
+	}
+	sql = fmt.Sprintf("%s order by time DESC limit 1", sql)
```

Comment:
this query may return one of hostname data，we need return all hostname's new data, so we can change the query to 
```
select top(time, 1), reqActive, reqDurationNs, xxx from tableName group by hostname where time >= now() - 1m
```

Reply:
Okay, I've solved it.